monitoring/ping.go

```diff
@@ -0,0 +1,283 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = c.checkNodesRTT(nodes, client)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// checkNodesRTT implements the bulk of the logic by checking the ping time
+// between this node (self) and the other Serf Cluster member nodes
+func (c *pingChecker) checkNodesRTT(nodes []serf.Member, client *serf.RPCClient) error {
+	// ping each other node and fail in case the results are over a specified
+	// threshold
+	for _, node := range nodes {
+		// skip pinging self
+		if c.self.Addr.String() == node.Addr.String() {
+			continue
+		}
+
+		rttNanoSec, err := calculateRTT(client, c.self, node)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		latencies, err := c.saveLatencyStats(rttNanoSec, node)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		latencyHistogram, err := c.buildLatencyHistogram(node.Name, latencies)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		c.logger.Debugf("%s <-ping-> %s = %dns [latest]", c.self.Name, node.Name, rttNanoSec)
+		c.logger.Debugf("%s <-ping-> %s = %dns [%.2f percentile]",
+			c.self.Name, node.Name,
+			latencyHistogram.ValueAtQuantile(latencyQuantile),
+			latencyQuantile)
+
+		latencyPercentile := latencyHistogram.ValueAtQuantile(latencyQuantile)
+		if latencyPercentile >= latencyThreshold.Nanoseconds() {
+			c.logger.Warningf("%s <-ping-> %s = slow ping detected. Value %dns over threshold %s (%dns)",
+				c.self.Name, node.Name, latencyPercentile,
+				latencyThreshold.String(), latencyThreshold.Nanoseconds())
+		} else {
+			c.logger.Debugf("%s <-ping-> %s = ping okay. Value %dns within threshold %s (%dns)",
+				c.self.Name, node.Name, latencyPercentile,
+				latencyThreshold.String(), latencyThreshold.Nanoseconds())
+		}
+	}
+
+	return nil
+}
+
+// buildLatencyHistogram maps latencies to a HDRHistrogram
+func (c *pingChecker) buildLatencyHistogram(nodeName string, latencies []int64) (latencyHDR *hdrhistogram.Histogram, err error) {
+	latencyHDR = hdrhistogram.New(pingMinimum.Nanoseconds(),
+		pingMaximum.Nanoseconds(), pingSignificantFigures)
+
+	for _, v := range latencies {
+		err := latencyHDR.RecordValue(v)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	return latencyHDR, nil
+}
+
+// saveLatencyStats is used to store ping values in HDR Histograms in memory
+func (c *pingChecker) saveLatencyStats(pingLatency int64, node serf.Member) (latencies []int64, err error) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	if value, exists := c.latencyStats.Get(node.Name); exists {
+		var ok bool
+		if latencies, ok = value.([]int64); !ok {
+			return nil, trace.BadParameter("couldn't parse node latency as []int64 on %s", c.serfMemberName)
+		}
+	}
+
+	if len(latencies) >= latencyStatsSlidingWindowSize {
+		// keep the slice within the sliding window size
+		// slidingWindowSize is -1 because another element will be added a few lines below
+		latencies = latencies[1 : latencyStatsSlidingWindowSize-1]
```

Comment:
this will override the last new value. The range is not right-inclusive - the last remaining element will be `latencies[latencyStatsSlidingWindowSize-2]`. You need to:
```go
latencies = latencies[1 : latencyStatsSlidingWindowSize]
```
nit: also, in the suggested version the slice reused the same backing array which stayed `latencyStatsSlidingWindowSize` in size - this implementation will be constantly reallocating (`latencyStatsSlidingWindowSize * 2` in worst-case).